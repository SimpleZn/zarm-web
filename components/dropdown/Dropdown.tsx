import React, { KeyboardEventHandler } from 'react';
import { Popper } from 'zarm';
import classnames from 'classnames';
import { DropdownProps } from './PropsType';

const defaultProps = {
  prefixCls: 'zw-dropdown',
  direction: 'bottomLeft',
  trigger: 'click',
  disabled: false,
  shape: 'radius',
};

interface DropdownStates {
  visible: boolean;
}

function getVisible(props, state) {
  const { visible: propsState } = props;
  const { visible: stateProps } = state;
  return propsState === undefined ? stateProps : propsState;
}

export default class Dropdown extends React.Component<DropdownProps, DropdownStates> {
  static defaultProps = defaultProps;

  static visibleList = new Set();

  triggerPointRef = React.createRef<HTMLSpanElement>();

  popperContentRef = React.createRef<HTMLDivElement>();

  private hoverTimer?: number;

  private prevActiveElem?: Element;

  state: DropdownStates = {
    visible: false,
  };

  componentDidMount() {
    document.addEventListener('mousedown', this.onDocClick);
  }

  componentDidUpdate() {
    this.prevActiveElem = document.activeElement || document.body;
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.onDocClick);
  }

  onDocClick = (e: MouseEvent) => {
    const visible = getVisible(this.props, this.state);
    const { current } = this.triggerPointRef;
    const { current: popperContentCurrent } = this.popperContentRef;
    if (visible) {
      if (current && popperContentCurrent) {
        if (popperContentCurrent.contains(e.target as Node) || current.contains(e.target as Node)) {
          return;
        }
      }
      this.onVisibleChange(false);
    }
  };

  onVisibleChange = (visible: boolean) => {
    const { disabled, onVisibleChange, visible: propsVisible } = this.props;
    if (disabled) {
      return;
    }
    if (visible) {
      Dropdown.visibleList.add(this);
    } else {
      Dropdown.visibleList.delete(this);
    }
    if (propsVisible === undefined) {
      return this.setState({
        visible,
      });
    }
    if (typeof onVisibleChange === 'function') {
      onVisibleChange(visible);
    }
  };

  onClick = () => {
    const { trigger } = this.props;
    const visible = getVisible(this.props, this.state);
    if (trigger !== 'click') {
      return;
    }
    this.onVisibleChange(!visible);
  };

  onMouseEnter = () => {
    const { trigger, visible } = this.props;
    if (trigger !== 'hover') {
      return;
    }
    if (this.hoverTimer) {
      clearTimeout(this.hoverTimer);
    }
    if (!visible) {
      this.onVisibleChange(true);
    }
  };

  onMouseLeave = () => {
    const { trigger } = this.props;
    const visible = getVisible(this.props, this.state);
    if (trigger !== 'hover' || !visible) {
      return;
    }
    this.hoverTimer = setTimeout(() => {
      this.onVisibleChange(false);
    }, 300);
  };

  onContextMenu = (e: React.MouseEvent<HTMLSpanElement>) => {
    const { trigger } = this.props;
    const visible = getVisible(this.props, this.state);
    if (trigger !== 'contextMenu') {
      return;
    }
    e.preventDefault();
    this.onVisibleChange(!visible);
  };

  onKeydown: KeyboardEventHandler = (e) => {
    const visible = getVisible(this.props, this.state);
    if (visible && e.keyCode === 27) {
      e.stopPropagation();
      if (this.popperContentRef.current && this.popperContentRef.current.contains(e.currentTarget)) {
        const { prevActiveElem } = this;
        if (prevActiveElem && prevActiveElem instanceof HTMLElement) {
          setTimeout(() => {
            prevActiveElem.focus();
          });
        }
      }
      this.onVisibleChange(false);
    }
  };

  render() {
    const {
      children,
      className,
      prefixCls,
      onVisibleChange,
      disabled,
      content,
      trigger,
      shape,
      ...others
    } = this.props;
    const visible = getVisible(this.props, this.state);
    const cls = classnames(prefixCls, className, `${prefixCls}--${shape}`);
    const dropdownContent = (
      <div
        tabIndex={-1}
        ref={this.popperContentRef}
        onMouseEnter={this.onMouseEnter}
        onMouseLeave={this.onMouseLeave}
        onKeyDown={this.onKeydown}
      >
        {content}
      </div>
    );

    return (
      <Popper
        {...others}
        visible={disabled ? false : visible}
        className={cls}
        trigger="manual"
        onVisibleChange={this.onVisibleChange}
        content={dropdownContent}
      >
        <span
          ref={this.triggerPointRef}
          style={{ display: 'inline-block' }}
          onClick={this.onClick}
          onMouseEnter={this.onMouseEnter}
          onMouseLeave={this.onMouseLeave}
          onContextMenu={this.onContextMenu}
          onKeyDown={this.onKeydown}
        >
          {children}
        </span>
      </Popper>
    );
  }
}
