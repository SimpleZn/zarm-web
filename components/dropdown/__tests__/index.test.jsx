import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Dropdown from '../index';
import Button from '../../button';

describe('Dropdown', () => {
  it('renders normal Dropdown correctly', () => {
    const doneChange = jest.fn();
    const wrapper = render(
      <div>
        <Dropdown
          onVisibleChange={doneChange}
        >
          <Button>click</Button>
        </Dropdown>
      </div>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  const timer = function (t) {
    return new Promise((resovle) => {
      setTimeout(() => {
        resovle(true);
      }, t);
    });
  };

  it('trigger onVisibleChange correctly', async () => {
    let visible = false;
    const doneChange = jest.fn();
    const onClick = jest.fn();
    const wrapper = mount(
      <Dropdown
        visible={visible}
        onVisibleChange={(e) => {
          visible = e;
          doneChange(e);
        }}
      >
        <Button className="btn" onClick={onClick}>click</Button>
      </Dropdown>,
    );
    wrapper.find('button').simulate('click');
    expect(onClick).toHaveBeenCalled();
    expect(doneChange).toHaveBeenCalled();
    expect(visible).toEqual(true);
  });

  it('not trigger onVisibleChange when trigger is not click', async () => {
    let visible = false;
    const doneChange = jest.fn();
    const onClick = jest.fn();
    const wrapper = mount(
      <Dropdown
        visible={visible}
        trigger="hover"
        onVisibleChange={(e) => {
          visible = e;
          doneChange(e);
        }}
      >
        <Button className="btn" onClick={onClick}>click</Button>
      </Dropdown>,
    );
    wrapper.find('button').simulate('click');
    expect(onClick).toHaveBeenCalled();
    expect(doneChange).not.toHaveBeenCalled();
    expect(visible).toEqual(false);
  });


  it('trigger onVisibleChange correctly when props trigger is hover', async () => {
    let visible = false;
    const doneChange = jest.fn();
    const wrapper = mount(
      <Dropdown
        visible={visible}
        trigger="hover"
        onVisibleChange={(e) => {
          visible = e;
          doneChange(e);
        }}
      >
        <Button className="btn">hover</Button>
      </Dropdown>,
    );
    wrapper.find('span').first().simulate('mouseEnter');
    expect(doneChange).toHaveBeenCalled();
    expect(visible).toEqual(true);
  });


  it('not trigger onVisibleChange correctly when props trigger is not hover', async () => {
    let visible = false;
    const doneChange = jest.fn();
    const wrapper = mount(
      <Dropdown
        visible={visible}
        trigger="click"
        onVisibleChange={(e) => {
          visible = e;
          doneChange(e);
        }}
      >
        <Button className="btn">hover</Button>
      </Dropdown>,
    );
    wrapper.find('span').first().simulate('mouseEnter');
    expect(doneChange).not.toHaveBeenCalled();
    expect(visible).toEqual(false);
  });


  it('trigger onVisibleChange correctly when props trigger is mouseleave', async () => {
    let visible = true;
    const doneChange = jest.fn();
    const wrapper = mount(
      <Dropdown
        visible={visible}
        trigger="hover"
        onVisibleChange={(e) => {
          visible = e;
          doneChange(e);
        }}
      >
        <Button className="btn">hover</Button>
      </Dropdown>,
    );
    const span = wrapper.find('span').first();
    span.simulate('mouseLeave');
    await timer(310);   // 实际实现中会有300ms的延迟才消失
    expect(doneChange).toHaveBeenCalled();
    expect(visible).toEqual(false);
  });


  it('not trigger onVisibleChange correctly when props trigger is mouseleave and visible is false', async () => {
    let visible = false;
    const doneChange = jest.fn();
    const wrapper = mount(
      <Dropdown
        visible={visible}
        trigger="hover"
        onVisibleChange={(e) => {
          visible = e;
          doneChange(e);
        }}
      >
        <Button className="btn">hover</Button>
      </Dropdown>,
    );
    const span = wrapper.find('span').first();
    span.simulate('mouseLeave');
    await timer(310);   // 实际实现中会有300ms的延迟才消失
    expect(doneChange).not.toHaveBeenCalled();
    expect(visible).toEqual(false);
  });

  // 当鼠标离开不超过300ms的时候
  it('not trigger onVisibleChange correctly when props trigger is mouseleave', async () => {
    let visible = true;
    const doneChange = jest.fn();
    const wrapper = mount(
      <Dropdown
        visible={visible}
        trigger="hover"
        onVisibleChange={(e) => {
          visible = e;
          doneChange(e);
        }}
      >
        <Button className="btn">hover</Button>
      </Dropdown>,
    );
    const span = wrapper.find('span').first();
    span.simulate('mouseLeave');
    await timer(100);   // 实际实现中会有300ms的延迟才消失
    expect(doneChange).not.toHaveBeenCalled();
    expect(visible).toEqual(true);
  });


  it('renders normal Dropdown correctly when trigger contextmenu', async () => {
    let visible = false;
    const doneChange = jest.fn();
    const wrapper = mount(
      <Dropdown
        visible={visible}
        trigger="contextMenu"
        onVisibleChange={(e) => {
          visible = e;
          doneChange(e);
        }}
      >
        <Button> contextmenu </Button>
      </Dropdown>,
    );
    wrapper.find('button').simulate('contextmenu');
    await timer(100);
    expect(doneChange).toHaveBeenCalled();
    expect(visible).toEqual(true);
  });

  it('should not trigger onVisibleChange callback when trigger is not contextMenu', async () => {
    let visible = false;
    const doneChange = jest.fn();
    const wrapper = mount(
      <Dropdown
        visible={visible}
        trigger="click"
        onVisibleChange={(e) => {
          visible = e;
          doneChange(e);
        }}
      >
        <Button>contextmenu</Button>
      </Dropdown>,
    );
    wrapper.find('button').simulate('contextmenu');
    await timer(100);
    expect(doneChange).not.toHaveBeenCalled();
    expect(visible).toEqual(false);
  });

  it('should not trigger onVisibleChange callback when disabled', async () => {
    let visible = false;
    const doneChange = jest.fn();
    const wrapper = mount(
      <Dropdown
        disabled
        visible={visible}
        trigger="click"
        onVisibleChange={(e) => {
          visible = e;
          doneChange(e);
        }}
      >
        <Button>contextmenu</Button>
      </Dropdown>,
    );
    wrapper.find('button').simulate('click');
    expect(doneChange).not.toHaveBeenCalled();
  });
});
