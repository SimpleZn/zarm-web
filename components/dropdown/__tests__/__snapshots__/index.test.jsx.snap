// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders normal Dropdown correctly 1`] = `
<div>
  <span
    style="display:inline-block"
  >
    <button
      class="zw-button zw-button--default zw-button--md zw-button--radius"
      type="button"
    >
      <span>
        click
      </span>
    </button>
  </span>
</div>
`;
